### DISCUSSION TITLE
Support for OpenID Federation 1.0

### DISCUSSION THREAD

**Discussion by thomasdarimont** [keycloak/keycloak#31027] [Ideas]
The following discussion briefly explains what OpenID Federation is about and asks the community to provide some feedback on the following question: "How would you benefit from having OpenID Federation support in Keycloak?"

# OpenID Federation

The OpenID Federation (OIDFED) protocol, which Keycloak now supports, defines a highly scalable architecture for establishing trust between multiple parties based on JWT/JOSE artifacts, ensuring that your system can grow with your needs.

At a high level, this protocol enables the following use cases:

- Determination whether an OpenID relying party (RP) and an OpenID provider (OP) can trust one another, given one or more intermediaries / shared trust anchors;
- Standardized way for discovery, policing, and updating of RP and OP metadata, federation-wide
- A standardized way for registration of RPs as clients at OPs, saving costly human administration.

Even though the current [OpenID Federation 1.0 standard document](https://openid.net/specs/openid-federation-1_0.html) is still a draft (draft 36), it is under active development and is already used by large institutions and some governments.

According to [this presentation by one of the spec authors](https://self-issued.info/presentations/OpenID_Federation_1.0_for_18th_FIM4R_Workshop_at_TIIME_2024.pdf):
The core components have been stable since draft 20 (now at draft 36). Useful (optional) backward-compatible features have been added. The Last Implementer’s Draft is imminent, and the final specification is anticipated for later this year.

# Discussions on OpenID Federation support and Keycloak

In the past, there was some community interest in Keycloak supporting the OpenID Federation standard.

- [KEYCLOAK-16574 - OpenID Connect Federation Community proposal](https://github.com/keycloak/keycloak-community/pull/201/files)
- [OIDC federation as a relying party](https://github.com/keycloak/keycloak/discussions/26436)
- [Keycloak as Relying Party in OpenID Connect Federation 1.0 - draft 29](https://github.com/keycloak/keycloak/discussions/23435)

# Existing PoCs
- [An early PoC implementation based on an old OIDFED draft](https://github.com/eosc-kc/keycloak-oidc-federation) and an [updated version based on Keycloak 25](https://github.com/thomasdarimont/keycloak-oidc-federation/tree/update/keycloak-25.0.x)

# Additional links
- [Example for OpenID Federation API](https://app.swaggerhub.com/apis/SphereonInt/OpenIDFederationAPI/1.0.0-d35)
- [OpenID Federation explained Takahiko Kawasaki / Authlete](https://darutk.medium.com/identity-federation-by-openid-connect-9f8f0a189826)
- [OpenID Federation explained by connect2id](https://connect2id.com/learn/openid-federation)

**Comment by cgeorgilakis** [comment]
I am part of the team originally implemented [keycloak-oidc-federation](https://github.com/eosc-kc/keycloak-oidc-federation). It is a previous work based in previous draft (13) and previous Keycloak (version 11). The work has been abandoned and is not used in any of our Keycloak instances. Our implementation is only for explicit registration. Our [Keycloak community PR]( https://github.com/keycloak/keycloak-community/pull/201) is also for automatic configuration.

Our initial plan is to start implementing it again from October. We want to support it for European Digital Identity Framework and [eduGAIN profile for OpenID Federation]( https://github.com/GEANT/edugain-openidfed/blob/main/OpenID4RS-OpenIDFederation.md) in our Research & Education Keycloak installation. Thanks @thomasdarimont for your work in this and updating our extension. We believe that Keycloak should try supporting OpenID Federation. We could start with some features as an experimental feature in Keycloak core. We could discuss a collaboration/ help for it based on availability.

**Comment by francis-pouatcha** [comment]
We will leverage OpenID Federation to enable **Automatic Client Registration** within the context of the European [FIDA](https://www.europeanpaymentscouncil.eu/news-insights/insight/exploring-fida-revised-rules-financial-data-sharing-access) and also backport our implementation of the European [PSD2/PSD3/PSR](https://www.europeanpaymentscouncil.eu/news-insights/insight/what-do-psd3-and-psr-mean-payments-sector) regulations. 

In mainland Europe, proprietary protocols like the Berlin Group [Open Finance - NextGenPSD2 API Framework](https://www.berlin-group.org/open-finance) currently utilize [eIDAS](https://eidas.ec.europa.eu/efda/browse/notification/eid-chapter-contacts) certificates issued by accredited certification authorities to build trust. Combining these eIDAS certificates with [Entity Statement](https://openid.net/specs/openid-federation-1_0.html#name-entity-statement) can provide a robust mechanism for **Automatic Registration of Clients** within the European PSD/PSR/FIDA ecosystem.

**Reply by francis-pouatcha** [reply]
Just for information from today's breakout session: https://www.ietf.org/archive/id/draft-parecki-oauth-client-id-metadata-document-01.html

**Comment by bucchi** [comment]
In these couple of months, I have started looking into OpenID Federation spec  for implementation with Keycloak since I found out this interesting proposal:
https://peppelinux.github.io/federation-wallet/main.html


Right nowI, I am working on the @thomasdarimont ’s branch ([updated version based on Keycloak 25](https://github.com/thomasdarimont/keycloak-oidc-federation/tree/update/keycloak-25.0.x)) to deploy in my local environment.
 
The following is the question about the deployment(, which might better  be asked not here , but at other place. If so, let me know.)


[Question]
As usual for custom providers, I added service configuration files in META-INF/services into the cloned branch. Building that with default pom.xml and deploying that, I have an error below for “kc.sh start-dev” after “kc.sh build”. 
If you have any insight about this, please let me know.


```
2024-09-16 07:16:30,191 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (development) mode
2024-09-16 07:16:30,191 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) Error details:: java.lang.NullPointerException: Cannot invoke "jakarta.persistence.EntityManagerFactory.createEntityManager(jakarta.persistence.SynchronizationType)" because "emf" is null
	at org.keycloak.quarkus.runtime.storage.database.jpa.AbstractJpaConnectionProviderFactory.createEntityManager(AbstractJpaConnectionProviderFactory.java:111)
	at org.keycloak.quarkus.runtime.storage.legacy.database.QuarkusJpaConnectionProviderFactory.create(QuarkusJpaConnectionProviderFactory.java:77)
	at org.keycloak.quarkus.runtime.storage.legacy.database.QuarkusJpaConnectionProviderFactory.create(QuarkusJpaConnectionProviderFactory.java:62)
	at org.keycloak.services.DefaultKeycloakSession.getProvider(DefaultKeycloakSession.java:177)
	at org.keycloak.models.jpa.JpaRealmProviderFactory.create(JpaRealmProviderFactory.java:64)
	at org.keycloak.models.jpa.JpaRealmProviderFactory.create(JpaRealmProviderFactory.java:40)
	at org.keycloak.services.DefaultKeycloakSession.getProvider(DefaultKeycloakSession.java:177)
	at org.keycloak.models.cache.infinispan.RealmCacheSession.getRealmDelegate(RealmCacheSession.java:147)
	at org.keycloak.models.cache.infinispan.RealmCacheSession.getRealmsStream(RealmCacheSession.java:497)
	at org.keycloak.protocol.oidc.federation.op.tasks.DefaultClientExpiryTasksFactory.postInit(DefaultClientExpiryTasksFactory.java:28)
	at org.keycloak.quarkus.runtime.integration.QuarkusKeycloakSessionFactory.init(QuarkusKeycloakSessionFactory.java:105)
	at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.createSessionFactory(QuarkusKeycloakApplication.java:64)
	at org.keycloak.services.resources.KeycloakApplication.startup(KeycloakApplication.java:91)
```

**Reply by DobryninIlya** [reply]
did you find the reason for the error?

**Comment by OrangeApricorn** [comment]
Hi, we were using your project to try to customise our own IdP/OIDC. 
While we were trying to persist any data into the PostgreSQL table "oidc_federation_config" referred in the OIDCFedConfigEntity class, via POST API: ` http://localhost:8080/realms/my-realm/oidc-federation/configuration` no data are persisted and we get: 
`404 NOT FOUND   "errorMessage": "Failed to create configuration for the realm my-realm"`

Is the POST API correctly used? Are there any known issues about this problem while persisting?


This is the JSON Body used in the API:
```
{
  "authorityHints": [
    "http://localhost:8080/realms/my-iam/protocol/openid-connect/userinfo"
  ],
  "expirationTime": 3600,
  "trustAnchors": [
    "http://localhost:8080/realms/my-iam/protocol/openid-connect/userinfo"
  ],
  "registrationType": "string"
}
```


This is the detailed StackTrace:

```
ERROR [org.keycloak.protocol.oidc.federation.op.model.OIDCFedConfigServiceImpl] (executor-thread-8) Failed to retrieve OIDCFedConfigEntity for realm eb0532eb-4575-4406-94bd-6567f384851b: org.keycloak.models.ModelException: Database operation failed
keycloak-1  |   at org.keycloak.connections.jpa.PersistenceExceptionConverter.convert(PersistenceExceptionConverter.java:109)
keycloak-1  |   at org.keycloak.connections.jpa.PersistenceExceptionConverter.invoke(PersistenceExceptionConverter.java:68)
keycloak-1  |   at jdk.proxy2/jdk.proxy2.$Proxy61.find(Unknown Source)
keycloak-1  |   at org.keycloak.protocol.oidc.federation.op.model.OIDCFedConfigServiceImpl.getEntity(OIDCFedConfigServiceImpl.java:31)
keycloak-1  |   at org.keycloak.protocol.oidc.federation.op.rest.FederationConfigurationEndpoint.saveConfiguration(FederationConfigurationEndpoint.java:100)
keycloak-1  |   at org.keycloak.protocol.oidc.federation.op.rest.FederationConfigurationEndpoint$quarkusrestinvoker$saveConfiguration_0ce599a9201118b200111aa3048c10fcccf9423c.invoke(Unknown Source)
keycloak-1  |   at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
keycloak-1  |   at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:141)
keycloak-1  |   at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
keycloak-1  |   at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:635)
keycloak-1  |   at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
keycloak-1  |   at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
keycloak-1  |   at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
keycloak-1  |   at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
keycloak-1  |   at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
keycloak-1  |   at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
keycloak-1  |   at java.base/java.lang.Thread.run(Thread.java:1583)
keycloak-1  | Caused by: java.lang.IllegalArgumentException: Unable to locate entity descriptor: org.keycloak.protocol.oidc.federation.op.model.OIDCFedConfigEntity
keycloak-1  |   at org.hibernate.internal.SessionImpl.find(SessionImpl.java:2494)
keycloak-1  |   at org.hibernate.internal.SessionImpl.find(SessionImpl.java:2425)
keycloak-1  |   at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
keycloak-1  |   at java.base/java.lang.reflect.Method.invoke(Method.java:580)
keycloak-1  |   at org.keycloak.connections.jpa.PersistenceExceptionConverter.invoke(PersistenceExceptionConverter.java:66)
keycloak-1  |   ... 15 more
keycloak-1  | Caused by: org.hibernate.UnknownEntityTypeException: Unable to locate entity descriptor: org.keycloak.protocol.oidc.federation.op.model.OIDCFedConfigEntity
keycloak-1  |   at org.hibernate.metamodel.model.domain.internal.MappingMetamodelImpl.getEntityDescriptor(MappingMetamodelImpl.java:439)
keycloak-1  |   at org.hibernate.internal.SessionImpl.requireEntityPersister(SessionImpl.java:2329)
keycloak-1  |   at org.hibernate.internal.SessionImpl.byId(SessionImpl.java:1171)
keycloak-1  |   at org.hibernate.internal.SessionImpl.find(SessionImpl.java:2456)
keycloak-1  |   ... 19 more
```

We are developing under the Docker distribution, with PostgreSQL:16 and Quarkus

Thanks in advance!

**Reply by bucchi** [reply]
I have just created a [PR](https://github.com/thomasdarimont/keycloak-oidc-federation/pull/1)  for  [updated version based on Keycloak 25](https://github.com/thomasdarimont/keycloak-oidc-federation/tree/update/keycloak-25.0.x).

And I guess that the pull request could resolve your issue too.
So please try that, and let me know what happens.

**Comment by bucchi** [comment]
Some update information related to OpenID Federation;

- My article about "OpenID Federation with Keycloak" has been published recently on Medium.
https://bucchi.medium.com/building-trust-with-openid-federation-trust-chain-on-keycloak-f8ac021add3a

- OpenID Federation Interop Event coming soon
https://openid.net/openid-federation-interop-apr-28-30-2025/

- Digital Identity Wallet use case for OpenID Federation
https://tac.openwallet.foundation/projects/openid-federation-ts/?h=federation
https://openid.net/specs/openid-federation-wallet-1_0.html

**Reply by cgeorgilakis** [reply]
> Some update information related to OpenID Federation;
> 
> * My article about "OpenID Federation with Keycloak" has been published recently on Medium.
>   https://bucchi.medium.com/building-trust-with-openid-federation-trust-chain-on-keycloak-f8ac021add3a

@bucchi Do you know if plugin bypass OIDF Federation tests for explicit registration?

> * OpenID Federation Interop Event coming soon
>   https://openid.net/openid-federation-interop-apr-28-30-2025/

Very interesting event.
The problem is that it is too soon based on work that can be done before. Is anyone else from Keycloak team interested to participate?

> * Digital Identity Wallet use case for OpenID Federation
>   https://tac.openwallet.foundation/projects/openid-federation-ts/?h=federation
>   https://openid.net/specs/openid-federation-wallet-1_0.html

**Reply by bucchi** [reply]
I agree with you.

Keycloak was not ready for that event at that time and I guess nobody from Keycloak participated in the event.

But someone who participated in the event from a company called radiam ( https://www.raidiam.com/ ), gave me some messages in my linkedin account after reading my article on Medium.com.

And they asked me if I am interested in integrating Keycloak with their Trust Anchor service of OpenID Federation.

Is it pretty interesting to you guys, also? Let us discuss how we can work on about that.

**Reply by cgeorgilakis** [reply]
> I agree with you.
> 
> Keycloak was not ready for that event at that time and I guess nobody from Keycloak participated in the event.
> 
> But someone who participated in the event from a company called radiam ( https://www.raidiam.com/ ), gave me some messages in my linkedin account after reading my article on Medium.com.
> 
> And they asked me if I am interested in integrating Keycloak with their Trust Anchor service of OpenID Federation.
> 
> Is it pretty interesting to you guys, also? Let us discuss how we can work on about that.

We have create a dummy OpenId Federation with docker compose from .,https://github.com/go-oidfed/lib/tree/main/examples/edugain-pilot. I could provide details if needed.
Now, I develop and we are in test process of OpenID Federation in our Keycloak core.
Although you could try, you should take into account that OpenIdFederation extension produce wrong .well-known/openid-federation  data (fe typ and content-type must be 'entity-statement+jwt', jwks is not correct in all the case, wrong optional fields)

**Comment by eduperottoni** [comment]
Hey, guys!

The [implementation for Keycloak 26](https://github.com/thomasdarimont/keycloak-oidc-federation/tree/update/keycloak-26.x)'s documentation says that it's based on draft 13 and checking the code, it indeed doesn't support features like _automatic registration_. [OpenID Federation draft](https://openid.net/specs/openid-federation-1_0.html) has evolved quickly and is now on its 43rd version.

Searching on the web, I didn't find any other implementation of `OIDFed` using `Keycloak`. Is there anyone currently working on updating this current implementation to support all protocol's features and to be updated with the specification? Is this on the `Keycloak`s backlog?

**Comment by cgeorgilakis** [comment]
@eduperottoni I agree.

@eduperottoni @thomasdarimont @bucchi I have created [jira epic](https://github.com/keycloak/keycloak/issues/40509) for the need of implementaion of [Openid Federation](https://openid.net/specs/openid-federation-1_0.html).
I have started implementing [OpenID Federation OP with explicit registration](https://github.com/keycloak/keycloak/issues/40511) and I will make a PR when we are ready.
It will be nice if other teams could help in code or business logic. We could keep this discussion for general progress and discuss details in issues and future PRs.

**Reply by eduperottoni** [reply]
Great, @cgeorgilakis! Thanks for sharing!

We can help in the implementation, since we are interested in `Keycloak` working on a `OID Federation`. I'll start some discussions in the issue.

**Reply by bucchi** [reply]
@cgeorgilakis, I love your proposal.

I am in, too!!

BTW, as I posted here ( https://github.com/keycloak/keycloak/discussions/31027#discussioncomment-13498101 ), I got some proposal for integrating Keycloak with the real Trust Anchor service of OpenID Federation out there.

I guess it could be a pretty good target for us.
What do you think about that?

**Comment by cgeorgilakis** [comment]
For having tests for explicit client registration - as I guess the same issue is valid for automatic registration-, we need to moch the following REST API responses of Trust anchor:

- _/.well-known/openid-federation_ endpoint must return a valid jwt entity statement of Trust anchor
- In the returned jwt statement in the [5.1.1. ](https://openid.net/specs/openid-federation-1_0.html#section-5.1.1)[Federation Entity](https://openid.net/specs/openid-federation-1_0.html#name-federation-entity) the field _federation_fetch_endpoint_ contains a REST endpoint for quering information about RP. We need also to mock this response.

So, the question , mainly for Keycloak team @thomasdarimont, is how could we implement this two mock services? Should we use old or new test framework?

**Comment by cgeorgilakis** [comment]
I have implemented the first [PR](https://github.com/keycloak/keycloak/pull/41419) related to OpenID Federation as Keycloak being an OP with explicit registration.

Moreover, I will have the presentation **'Keycloak & OpenID Federation: Empowering Dynamic Trust in Federated Environments'** in [KeyConf25](https://keyconf.dev/) in 28 August in Amsterdam. See you there.

**Comment by stianst** [comment]
I think it would make a lot of sense to build on the concept of identity providers to provide this trust relationship rather than introducing a whole new thing. See https://github.com/keycloak/keycloak/discussions/42197

Also, look at https://github.com/keycloak/keycloak/pull/42229 (client assertions via 3rd party OIDC) or https://github.com/keycloak/keycloak/pull/42314 (client assertions via SPIFFE).

**Reply by cgeorgilakis** [reply]
> I think it would make a lot of sense to build on the concept of identity providers to provide this trust relationship rather than introducing a whole new thing. See #42197
> 
> Also, look at #42229 (client assertions via 3rd party OIDC) or #42314 (client assertions via SPIFFE).

The **OpenID Federation** and **SPIFFE** share similar goals while addressing different aspects of identity management. Both aim to enhance the trust and scalability of OIDC/OAuth ecosystems.

[OpenID Federation](https://openid.net/specs/openid-federation-1_0.html) is developed by the OpenID Connect Working Group, being nearing its Final Specification. It has gained strong interest among others from leading technology companies (fe authors, dedicated events). Most important implementation is [EU wallet ecosystem](https://ec.europa.eu/digital-building-blocks/sites/spaces/EUDIGITALIDENTITYWALLET/pages/694487738/EU+Digital+Identity+Wallet+Home#:~:text=The%20document%20sets%20out%20the%20common%20architecture%20of,the%20five%20initial%20implementing%20regulations%20adopted%20so%20far) ([OpenID Federation Wallet Architectures](https://openid.net/specs/openid-federation-wallet-1_0.html)), which supports linking national wallets, identity providers, and relying parties. It is combined with OpenID4VCI that Keycloak aims to preview it.
Countries such as Italy already use OpenID Federation. Large-scale federations like EduGAIN and the European Open Science Cloud (EOSC), including hundreds of Keycloak instances, are aimed to be created.
The OAuth SIG Community Meeting also refers “Enabling Automatic Client Registration within the European Ecosystem (PSD, PSR, FIDA)”. @bucci  and other could possibly mentions their use cases.
Given this context, we believe Keycloak, as a leading open-source IAM solution, should support both OpenID Federation and SPIFFE.

I propose rebasing my [draft PR](https://github.com/keycloak/keycloak/pull/41419) to align with the ongoing SPIFFE-related work and @bucci’s test implementations, aiming to finalize it when possible. Alternatively, @bucci could create a dedicated PR for Automatic Client Registration, including only the relevant parts from my draft PR.

**Comment by cgeorgilakis** [comment]
[OpenID Federation](https://openid.net/specs/openid-federation-1_0.html) is Final specification from 17 February.
You could see [related post](https://self-issued.info/?p=2813) for more information.
I have made a [presentation](https://drive.google.com/file/d/1q1lbLL0M2iSk_c6lCYV_Fsnr01DNQ5-t) in Unconference / Open Space of Keycloak DevDay for my team work with OpenID Federation.

We will be glad to rebase my draft PR and proceed the talk for supporting OpenID Federation in Keycloak, if there is interest from Keycloak team. Full tests for OpenID Federation need more time and investigation as there is need to mock Trust Anchor and RP/OP responses. Any help is welcome.

**Comment by Dahkenangnon** [comment]
Regarding the challenge of mocking Trust Anchor endpoints for testing (.well-known/openid-federation and federation_fetch_endpoint), I'd like to point to @oidfed — a complete,  OpenID Federation 1.0 implementation for JavaScript/TypeScript, aligned with the Final specification (Feb 2026).

The @oidfed/authority package can spin up a spec-compliant Trust Anchor with both required endpoints wired out of the box, making it usable as a lightweight federation fixture in the Keycloak test suite — no handcrafted JWT mocks needed. It covers both explicit and automatic client registration via @oidfed/oidc.

For zero-setup exploration, the live explorer at https://explore.oidfed.com can resolve and inspect trust chains against any federation endpoint, including a Keycloak instance exposing /.well-known/openid-federation.





GitHub: https://github.com/Dahkenangnon/oidfed



Project home: https://oidfed.com/
